torch/jit/__init__.py

```diff
@@ -1743,6 +1743,32 @@ def _construct(cpp_module, init_fn):
             script_module._initializing = False
             return script_module
 
+        def _reconstruct(self, cpp_module):
+            """
+            Re-construct an instance of RecursiveScriptModule using an instance of a C++ module.
+
+            Arguments:
+                cpp_module: The C++ module that this RecursiveScriptModule will be rebuilt around.
+            """
+            self.__init__(cpp_module)
```

Comment:
do you need to delete existing attributes in order to restore the `self` to the state before `__init__`?

Reply:
I think the only attributes on the `ScriptModule` are:
1. memoized attributes for function calls to the underlying C++ module
2. ignored functions that run in Python 
3. other attributes marked for copying from the C++ to `ScriptModule` (internal API not used often)
4. attributes inherited from `nn.Module`

As discussed before, #2 and #3 are not recoverable from the C++ module (not a problem for your use case). There is code below to clear out all `ScriptFunctions` from `self.__dict__` which should take care of #3. 

@suo pls confirm.

Reply:
I think you are correct. `ScriptModule` is a Python wrapper that provides an `nn.Module`-like view over the C++ ScriptModule, so `reconstruct` should work by just swapping out the cpp module and updating the viewing machinery.

The place where this breaks down is places where, for debug purposes, `ScriptModule` is not a pure view over `cpp_module`, like `@ignore`. It might be good to error if we detect this cases.

Reply:
If I was writing this feature in isolation I would agree, but it looks like @jerryzh168 wants to keep those around? In his narrow use case, it looks like the pre- and post-reconstruction types are basically the same type so these `@ignore`d methods apply to the post-reconstruction type too?

Reply:
yeah, if it does not fit the correct behavior of what _reconstruct is supposed to do, maybe we can give it a more specific name, like `_reconstruct_same_instance`.